zetta_utils/training/lightning/train.py

```diff
@@ -84,55 +82,82 @@ def lightning_train(
     :param resource_limits: K8s reource limits per pod.
     :param resource_requests: K8s resource requests per pod.
     """
-
-    if local_run:
-        _lightning_train_local(regime, trainer, train_dataloader, val_dataloader=val_dataloader)
-        return
-
-    assert image is not None, "Must provide a container image for remote training."
-    assert resource_limits is not None, "Must provide resource limits for remote training."
-
-    execution_id = mazepa.id_generation.get_unique_id(
-        prefix="exec", slug_len=4, add_uuid=False, max_len=50
-    )
-
-    cluster_info = resource_allocation.k8s.parse_cluster_info(
-        cluster_name=cluster_name,
-        cluster_region=cluster_region,
-        cluster_project=cluster_project,
-    )
-
-    train_spec = {
-        "@type": "lightning_train",
-        "regime": builder.get_initial_builder_spec(regime),
-        "trainer": builder.get_initial_builder_spec(trainer),
-        "train_dataloader": builder.get_initial_builder_spec(train_dataloader),
-        "val_dataloader": builder.get_initial_builder_spec(val_dataloader),
-        "full_state_ckpt_path": full_state_ckpt_path,
+    args_mapping = {
+        "regime": regime,
+        "trainer": trainer,
+        "train_dataloader": train_dataloader,
+        "val_dataloader": val_dataloader,
     }
 
-    for _key in ["regime", "trainer", "train_dataloader"]:
-        assert train_spec[_key] is not None, f"{_key} requires builder compatible spec."
-
-    _lightning_train_remote(
-        execution_id,
-        cluster_info=cluster_info,
-        image=image,
-        num_nodes=num_nodes,
-        nproc_per_node=nproc_per_node,
-        retry_count=retry_count,
-        train_spec=train_spec,
-        env_vars=env_vars,
-        follow_logs=follow_logs,
-        host_network=num_nodes > 1,
-        resource_limits=resource_limits,
-        resource_requests=resource_requests,
-    )
-
-
-@builder.register("multinode_train_launch")
+    if local_run:
+        _lightning_train_local(
+            regime=regime if not isinstance(regime, dict) else builder.build(regime),
+            trainer=trainer if not isinstance(trainer, dict) else builder.build(trainer),
+            train_dataloader=train_dataloader
+            if not isinstance(train_dataloader, dict)
+            else builder.build(train_dataloader),
+            val_dataloader=val_dataloader
+            if not isinstance(val_dataloader, dict)
+            else builder.build(val_dataloader),
+            full_state_ckpt_path=full_state_ckpt_path,
+        )
+    else:
```

Comment:
there are no statements after the else so we can return and remove this block, one less level of indentation